js/src/data/reducer.js

```diff
@@ -387,6 +392,52 @@ const reducer = ( state = DEFAULT_STATE, action ) => {
 			);
 		}
 
+		case TYPES.RECEIVE_MAPPING_RULES: {
+			const { rules, pagination } = action;
+			const newRulesState = [ ...state.mc.mapping.rules.items ];
+
+			const start = ( pagination.page - 1 ) * pagination.perPage;
+			const end = pagination.perPage;
+			newRulesState.splice( start, end, ...rules );
+
+			return chainState( state, 'mc.mapping.rules' )
+				.setIn( 'items', newRulesState )
+				.setIn( 'total', pagination.total )
+				.setIn( 'pages', pagination.pages )
+				.end();
+		}
+
+		case TYPES.UPSERT_MAPPING_RULE: {
+			const { rule } = action;
+			const newRulesState = [ ...state.mc.mapping.rules.items ];
+
+			const ruleIndex = newRulesState.findIndex(
+				( el ) => el.id === rule.id
+			);
+
+			if ( ruleIndex >= 0 ) {
+				newRulesState[ ruleIndex ] = rule;
+			} else {
+				newRulesState.push( rule );
+			}
+
+			return chainState( state, 'mc.mapping.rules' )
+				.setIn( 'items', newRulesState )
+				.setIn( 'total', newRulesState.length )
```

Comment:
When there are more than 10 mapping rules, e.g. 30 rules, the value of `total` state will be set to an incorrect value of 11 for a moment if only the first page is loaded.

https://user-images.githubusercontent.com/17420811/194299248-da3e0db5-509e-4225-b507-63439c4b7194.mp4